client/ayon_core/hosts/zbrush/api/pipeline.py

```diff
@@ -0,0 +1,559 @@
+        """
+        Returns:
+            Union[str, None]: Current project name.
+        """
+
+        return self.get_current_context().get("project_name")
+
+    def get_current_folder_path(self):
+        """
+        Returns:
+            Union[str, None]: Current asset name.
+        """
+
+        return self.get_current_context().get("folder_path")
+
+    def get_current_task_name(self):
+        """
+        Returns:
+            Union[str, None]: Current task name.
+        """
+
+        return self.get_current_context().get("task_name")
+
+    def get_current_context(self):
+        context = get_current_workfile_context()
+        if not context:
+            return get_global_context()
+        if "project_name" in context:
+            return context
+        # This is legacy way how context was stored
+        return {
+            "project_name": context.get("project"),
+            "folder_path": context.get("folder_path"),
+            "task_name": context.get("task")
+        }
+    # --- Workfile ---
+    def open_workfile(self, filepath):
+        open_file_zscript = ("""
+[IFreeze,
+[MemCreate, currentfile, 1000, 0]
+[VarSet, filename, "{filepath}"]
+[MemWriteString, currentfile, #filename, 0]
+[FileNameSetNext, #filename]
+[IKeyPress, 13, [IPress, File:Open:Open]]]
+[MemDelete, currentfile]
+]
+    """).format(filepath=filepath)
+        execute_zscript(open_file_zscript)
+        set_current_file(filepath=filepath)
+        return filepath
+
+    def save_workfile(self, filepath=None):
+        if not filepath:
+            filepath = self.get_current_workfile()
+        filepath = filepath.replace("\\", "/")
+        save_file_zscript = ("""
+[IFreeze,
+[MemCreate, currentfile, 1000, 0]
+[VarSet, filename, "{filepath}"]
+[MemWriteString, currentfile, #filename, 0]
+[FileNameSetNext, #filename]
+[IKeyPress, 13, [IPress, File:SaveAs:SaveAs]]]
+[MemDelete, currentfile]
+]
+""").format(filepath=filepath)
+        context = get_global_context()
+        save_current_workfile_context(context)
+        # move the json data to the files
+        # shutil.copy
+        copy_ayon_data(filepath)
+        set_current_file(filepath=filepath)
+        execute_zscript(save_file_zscript)
+        return filepath
+
+    def work_root(self, session):
+        return session["AYON_WORKDIR"]
+
+    def get_current_workfile(self):
+        project_name = get_current_context()["project_name"]
+        folder_path = get_current_context()["folder_path"]
+        task_name = get_current_context()["task_name"]
+        work_dir = get_workdir(project_name, folder_path, task_name)
+        txt_dir = os.path.join(
+            work_dir, ".zbrush_metadata").replace(
+                "\\", "/"
+        )
+        with open (f"{txt_dir}/current_file.txt", "r") as current_file:
+            content = str(current_file.read())
+            filepath = content.rstrip('\x00')
+            current_file.close()
+            return filepath
```

Comment:
Knowing _nothing_ of Zbrush - why can't we rely on:
```
[FileNameGetLastUsed]
```

Or:
```python
def get_current_workfile():
    temp_file = get_tempfile_path()
    script = """
// Get FileNameGetNext if any defined, else return FileNameGetLastUsed
[If, [FileNameHasNext], [VarSet, filename, [FileNameGetNext]], [VarSet, filename, FileNameGetLastUsed]]
// Return filename variable value file temp file
[MemWriteString, filename, #filename, 0]
[MemSaveToFile, filename, "{temp_file}", 0]
[MemDelete, filename]
"""
    execute_zscript(script)

    # Get returned value as written from the zscript
    with open(temp_file, "r") as f:
        return f.read()
```
Of course consider this psuedocode.

Reply:
Ah it seems the first returns ANY filename that was used - including e.g. file exports or reading scripts.

And even though the docs state that `[IGetTitle, Tool:Current Tool]` returns the full path to the tool - it doesn't seem like it.

Anyway, Zbrush itself seems to "remember" the current project just fine. If you use File > Save As the window title updates accordingly, then you can change tools, export objects, and click File > Save Next and it'll increment your project file just fine to a new version.

As such, the current zbrush project path _should_ be available somewhere?

Reply:
Ok, so Zbrush is a pain in the ass apparently in this area. However, I'd still argue that we should avoid trying to store the current file in a file in the working directory and instead store the current filepath on save in a scripting memory block of ZBrush, e.g. using something like:
```
// On save and on open to set current filepath
// Create a memory block to hold the filepath
[MemDelete, AYON_CurrentFilePath] // Delete if it already existed
[MemCreate, AYON_CurrentFilePath, [StrLength, "{workfile_path}"]]
[MemWriteString, AYON_CurrentFilePath, "{workfile_path}"]

// Get current workfile
// Note: File will not be written if no current file memory block exists
// Write the memory block to file (to get output from zscript to python)
[MemSaveToFile, AYON_CurrentFilePath, "{tmpfile}", 1]
```

Reply:
Here are some of the options I've tried: https://www.zbrushcentral.com/t/scripting-get-current-file-project-path/458261

Reply:
> Ok, so Zbrush is a pain in the ass apparently in this area. However, I'd still argue that we should avoid trying to store the current file in a file in the working directory and instead store the current filepath on save in a scripting memory block of ZBrush, e.g. using something like:
> 
> ```
> // On save and on open to set current filepath
> // Create a memory block to hold the filepath
> [MemDelete, AYON_CurrentFilePath] // Delete if it already existed
> [MemCreate, AYON_CurrentFilePath, [StrLength, "{workfile_path}"]]
> [MemWriteString, AYON_CurrentFilePath, "{workfile_path}"]
> 
> // Get current workfile
> // Note: File will not be written if no current file memory block exists
> // Write the memory block to file (to get output from zscript to python)
> [MemSaveToFile, AYON_CurrentFilePath, "{tmpfile}", 1]
> ```

I can try with this again but i tried this similar method in my fork repo(you can find the old commit), but for some reason it doesnt work when saving and open other workfiles.

Reply:
> but for some reason it doesnt work after saving and open other workfiles.

Interesting - did you try with memory blocks or vars? Because vars should be lost indeed with other script executions - and memory blocks should stick around.

Also, you might have hit the bug that the script execution actually didn't finish and you checked the 'written' file before the zscript finished?

Reply:
> > but for some reason it doesnt work after saving and open other workfiles.
> 
> Interesting - did you try with memory blocks or vars? Because vars should be lost indeed with other script executions - and memory blocks should stick around.
> 
> Also, you might have hit the bug that the script execution actually didn't finish and you checked the 'written' file before the zscript finished?

it doesnt hit bug it is just that  it doesnt save the data correctly in the temp file.

Reply:
It worked fine for me from a single session. However, the memory block must exist - otherwise it won't create/write the file. If the memory block is also the exact size fo the filepath there are no null bytes in the written file and the contents is _exactly_ the filepath so you can just read the file in full.